src/main/java/roomescape/reservationtime/ReservationTimeService.java

```
@@ -0,0 +1,77 @@
+    ) {
+        this.reservationTimeRepository = reservationTimeRepository;
+    }
+
+    public ReservationTimeResponse create(final ReservationTimeRequest request) {
+        validateDuplicateTime(request.startAt());
+
+        final ReservationTime reservationTime = new ReservationTime(request.startAt());
+        final Long id = reservationTimeRepository.save(reservationTime);
+        final ReservationTime savedReservationTime = reservationTimeRepository.findById(id);
+        return ReservationTimeResponse.from(savedReservationTime);
+    }
+
+    public List<ReservationTimeResponse> findAll() {
+        return reservationTimeRepository.findAll().stream()
+                .map(ReservationTimeResponse::from)
+                .toList();
+    }
+
+    public List<AvailableReservationTimeResponse> findAllAvailableTimes(final Long themeId, final LocalDate date) {
+        final List<ReservationTime> allTime = reservationTimeRepository.findAll();
+        final List<ReservationTime> allByThemeIdAndDate = reservationTimeRepository.findAllByThemeIdAndDate(themeId,
+                date);
+
+        final Set<ReservationTime> reservationTimesByThemeAndDate = new HashSet<>(allByThemeIdAndDate);
+        return allTime.stream()
+                .map(reservationTime ->
+                        AvailableReservationTimeResponse.from(
+                                reservationTime,
+                                reservationTimesByThemeAndDate.contains(reservationTime)
+                        )
+                )
+                .toList();
+    }
+
+    public void deleteById(final Long id) {
+        if (!reservationTimeRepository.existsById(id)) {
+            throw new NotFoundException("시간이 존재하지 않습니다.");
+        }
+
+        try {
+            reservationTimeRepository.deleteById(id);
+        } catch (final DataIntegrityViolationException e) {
+            throw new BadRequestException("예약 시간이 사용중입니다.");
+        }
+    }
```

데이터베이스에서 던지는 예외들을 그대로 execptionHandler에게 던지기 보다는 제가 제어할 수 있는 영역을 넓히고 싶었고 handler에서 repository에 대해서 모르게 하고 싶었습니다.

---

음... ReservationTime 테이블의 경우에는 간단한 테이블이다보니 `DataIntegrityViolationException`이 발생할만한 케이스가 적을 수 있을 것 같은데요. 

예를들어 `Reservation` 테이블의 경우 FK가 2개나 걸려있기 때문에 둘 중에 어떤 컬럼의 문제인지를 예외만 캐치해서는 확인학기가 조금 어렵지 않을까 싶어요. 🤔

개인적인 생각에서는 DB 제약사항 에러를 이렇게 Service단에서 일일이 다 캐치하게되면 서비스 로직이 불필요하게 커지게될 것 같다고 생각되는데요. 그래서 전역 예외 처리에서 `DataAccessException`를 처리해주는게 낫지 않을까라고 생각되기도하는 것 같아요. (사실상 이런 유즈케이스를 사용자에게 제공하지 않는 것이 근본적인 해결책이라 생각됩니다)

이 부분은 다른 크루들과 논의해보면 어떨까요? 💬
논의한 내용이나 혹은 논의를 통해서 도출해낸 부기의 생각을 코멘트로 공유해주시면 좋을 것 같아요. 🤗

---

크루들과 토론해보고, 내린 결론은 다음과 같습니다.
1. Fail Fast 원칙을 지키는 것이 코드의 흐름을 이해하기 쉬웠고 제가 작성한 try-catch는 어디서 어떤 예외가 발생해서 처리될지 service 계층만 보고는 이해하기 어려웠습니다.
2. try-catch로 원치 않는 예외를 잡을 가능성이 존재했습니다. 예외는 범위가 크기 때문에 내가 원하지 않는 이유로 던져진 예외를 잡아버려 버그를 찾기 힘들 수 있다.
3. 스프링이 던지는 예외는 stack trace가 깊어 성능이 다소 떨어질 수 있다. 우리가 검증을 한 후, 작업을 하면 이런 문제가 발생하지는 않는다.
4. 하지만, 쿼리가 추가로 발생하는 문제점이 있다.

결론적으로 현재 데이터베이스 검증을 거쳐야하는(외래키 제약조건) 로직들은 쿼리가 추가로 발생하더라도 많은 양의 쿼리가 필요하지 않고, 많은 양의 요청이 발생하지 않을 것으로 예상되기 때문에 (예약 생성, 예약시간 삭제, 테마 삭제는 실제 배포 환경에서도 그렇게 많은 요청이 발생하지 않을 것으로 예상) 쿼리가 추가로 발생하는 단점이 있지만 예외 상황을 명확하게 서비스에서 검증함으로써 코드의 명확성과 가독성에서 나오는 유지보수의 이점, stack trace가 깊지 않은 예외만 던지는데서 나오는 성능상의 이점이 더 크다고 판단하였습니다. 
- 특히 지금의 코드는 로직이 어떤 흐름으로 진행되는지 알기가 정말 어렵다고 생각합니다. (try-catch를 통해 흐름이 제어되고 있어, repository의 예외 상황을 예측하기 어려움)

코드에 반영하였습니다! !